src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,571 @@
+            Request,
+        >,
+        ConfigError,
+    > {
+        let s3_config = self.aws_s3.as_ref().expect("s3 config wasn't provided");
+
+        let s3_options = s3_config.options.clone();
+
+        match s3_options.storage_class {
+            Some(class @ S3StorageClass::DeepArchive) | Some(class @ S3StorageClass::Glacier) => {
+                return Err(ConfigError::UnsupportedStorageClass {
+                    storage_class: format!("{:?}", class),
+                });
+            }
+            _ => (),
+        }
+
+        let s3 = S3Sink { client };
+        let svc = ServiceBuilder::new()
+            .map(move |req| {
+                build_s3_request(req, bucket.clone(), prefix.clone(), s3_options.clone())
+            })
+            .settings(request, S3RetryLogic)
+            .service(s3);
+        Ok(svc)
+    }
+}
+
+fn build_s3_request(
+    req: PartitionInnerBuffer<Vec<u8>, Bytes>,
+    bucket: String,
+    path_prefix: Option<String>,
+    options: S3Options,
+) -> Request {
+    let (inner, key) = req.into_parts();
+
+    let filename = Uuid::new_v4().to_string();
+
+    let key = String::from_utf8_lossy(&key[..]).into_owned();
+    let key = format!(
+        "{}/{}{}.{}",
+        path_prefix.unwrap_or_default(),
+        key,
+        filename,
+        "json.gz"
+    )
+    .replace("//", "/");
+
+    debug!(
+        message = "Sending events.",
+        bytes = ?inner.len(),
+        bucket = ?bucket,
+        key = ?key
+    );
+
+    Request {
+        body: inner,
+        bucket,
+        key,
+        content_encoding: Compression::gzip_default().content_encoding(),
+        options: aws_s3::S3Options {
+            acl: options.acl,
+            grant_full_control: options.grant_full_control,
+            grant_read: options.grant_read,
+            grant_read_acp: options.grant_read_acp,
+            grant_write_acp: options.grant_write_acp,
+            server_side_encryption: options.server_side_encryption,
+            ssekms_key_id: options.ssekms_key_id,
+            storage_class: options.storage_class,
+            tags: options.tags,
+            content_encoding: None,
+            content_type: None,
+        },
+    }
+}
+
+/// Applies the following transformations to align event's schema with DD:
+/// - `_id` is generated in the sink(format described below);
+/// - `date` is set from the Global Log Schema's `timestamp` mapping, or to the current time if missing;
+/// - `message`,`host` are set from the corresponding Global Log Schema mappings;
+/// - `source`, `service`, `status`, `tags` and other reserved attributes are left as is;
+/// - the rest of the fields is moved to `attributes`.
+fn encode_event(event: Event) -> Option<EncodedEvent<PartitionInnerBuffer<Vec<u8>, Bytes>>> {
+    lazy_static! {
+        static ref KEY_PREFIX: Template =
+            Template::try_from("/dt=%Y%m%d/hour=%H/").expect("invalid object key format");
+    }
```

Comment:
Presumably `encode_event` will be in a hot loop. The use of `lazy_static` here follows up on my concerns expressed elsewhere. Please find a way to write this function without the use of `lazy_static`.

Reply:
Refactored this to get rid of `lazy_static`.